File: Tree/671. Second Minimum Node In a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findSecondMinimumValue = function(root) {
    const nodes = [];
    const stack = [root];
    while(stack.length) {
        const node = stack.pop();
        if (node) {
            nodes.push(node.val);
            if (node.val === root.val) {
                stack.push(node.right);
                stack.push(node.left);
            }
        }
    }
    let [min, result] = [root.val, Infinity];
    nodes.forEach(i => {
        if (min < i && i < result) {
          result = i; 
        }
    });
    return result !== Infinity ? result : -1;
};

File: Golang/Tree/671.second-minimum-node-in-a-binary-tree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findSecondMinimumValue(root *TreeNode) int {
	if root.Left != nil {
		stack := []*TreeNode{root.Left, root.Right}
		cursor := math.MaxInt32 + 1
		for len(stack) > 0 {
			node := stack[0]
			stack = stack[1:]
			if node.Val != root.Val && node.Val < cursor {
				cursor = node.Val
			}
			if node.Left != nil {
				stack = append(stack, node.Left, node.Right)
			}
		}
		if cursor != math.MaxInt32+1 {
			return cursor
		}
	}
	return -1
}
